bin/host/src/handler.rs

```diff
@@ -111,47 +132,121 @@ pub async fn fetch_eigenda_hint(
+            | HostHandlerError::UndefinedStatusCodeError(e) => {
+                return Err(anyhow!("failed to handle http response: {e}"))
+            }
+        }
+    } else {
+        // Handle success response
+        rollup_data = response
+            .bytes()
+            .await
+            .map_err(|e| anyhow!("should be able to get rollup payload from http response {e}"))?
+            .into();
+    }
+
+    Ok((is_valid, is_recent, rollup_data))
+}
+
+/// Store certificate validity in key-value store
+async fn store_cert_validity(
+    kv: SharedKeyValueStore,
+    altda_commitment: &AltDACommitment,
+    is_valid: bool,
+) -> Result<()> {
+    // Acquire a lock on the key-value store
+    let mut kv_write_lock = kv.write().await;
+    let mut validity_address = altda_commitment.digest_template();
     validity_address[RESERVED_EIGENDA_API_BYTE_INDEX] = RESERVED_EIGENDA_API_BYTE_FOR_VALIDITY;
 
     kv_write_lock.set(
         PreimageKey::new(*keccak256(validity_address), PreimageKeyType::GlobalGeneric).into(),
-        vec![claimed_validity as u8],
+        vec![is_valid as u8],
     )?;
 
-    // pre-populate eigenda blob field element by field element
-    let blob_length_fe = altda_commitment.get_num_field_element();
+    Ok(())
+}
 
+/// Store blob data in key-value store
+async fn store_blob_data(
+    kv: SharedKeyValueStore,
+    altda_commitment: &AltDACommitment,
+    rollup_data: Vec<u8>,
+) -> Result<()> {
+    // Acquire a lock on the key-value store
+    let mut kv_write_lock = kv.write().await;
+    // Prepare blob data
+    let blob_length_fe = altda_commitment.get_num_field_element();
     let eigenda_blob = EigenDABlobData::encode(rollup_data.as_ref(), PAYLOAD_ENCODING_VERSION_0);
 
-    // implementation requires eigenda_blob to be multiple of 32
+    // Verify blob data is properly formatted
     assert!(eigenda_blob.blob.len() % 32 == 0);
     let fetch_num_element = (eigenda_blob.blob.len() / BYTES_PER_FIELD_ELEMENT) as u64;
 
+    // Store each field element
     let mut field_element_key = altda_commitment.digest_template();
-    // populate every field element (fe) onto database
     for i in 0..blob_length_fe as u64 {
         field_element_key[72..].copy_from_slice(i.to_be_bytes().as_ref());
-
         let blob_key_hash = keccak256(field_element_key.as_ref());
 
         if i < fetch_num_element {
+            // Store actual blob data
             kv_write_lock.set(
                 PreimageKey::new(*blob_key_hash, PreimageKeyType::GlobalGeneric).into(),
                 eigenda_blob.blob[(i as usize) << 5..(i as usize + 1) << 5].to_vec(),
             )?;
         } else {
-            // empty bytes for the missing part between the re-encoded blob and claimed blob length from the header
+            // Fill remaining elements with zeros
             kv_write_lock.set(
                 PreimageKey::new(*blob_key_hash, PreimageKeyType::GlobalGeneric).into(),
```

Comment:
trying to understand this. why do we need to populate the entire blob_length? And not just the non-zero bytes? client could get the payload_size from the encoded_payload_header right? (not saying we need to make this change, just trying to reason/understand the logic)